---
title: "Filtering and QC for Seed companies"
author: "N.R Okereke"
date: "2025-01-24"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

***Load packages***

```{r}
library(dartR)
library(ape)
library(ade4)
library(tidyverse)
```

***Import data into genlight format***
```{r}
#This imports SNP mapping data and individual classification
gl <- gl.read.dart("../Report_DS24-10119_SNP_2.csv",
                   ind.metafile = "../STRAIT.KIN Metadata File.csv")
```
***Assign population and strata***
```{r}
pop(gl) <- gl@other$ind.metrics$Pop
strata(gl) <- data.frame(gl@other$ind.metrics$Generation)
nameStrata(gl) <- ~Generation
```


***Rename the sample names (individuals)***
```{r}
gl <- gl.recode.ind(gl, ind.recode = "../Rename_ALL.csv",
                    mono.rm = FALSE,
                    recalc = FALSE)
```



***Identify absent markers***
```{r}
chromV  <- gl@other$loc.metrics$Chrom_Sorghum_bicolor_v5.1
dropped <- gl$loc.names[grep("^Chr", as.character(chromV), invert = TRUE)]
```

***Filter out absent markers***

```{r}
Filter1 <- gl %>%
  gl.drop.loc(dropped) %>%
  gl.recalc.metrics()# drop missing data ---
```

***Examine genlight using report functions to determine appropriate threshold ***
```{r}
#determine call rate for loci and individual
gl.report.callrate(Filter1, method = "loc")
gl.report.callrate(Filter1, method = "ind")
#determine other parameter
gl.report.maf(Filter1)
gl.report.monomorphs(Filter1)
gl.report.reproducibility(Filter1)
gl.report.secondaries(Filter1)
```

***Filter using thresholds from reports***
```{r}
#filter genlight based on filtering criteria
filtered_gl <- gl %>%
  gl.filter.callrate(method = "loc", threshold = 0.672) %>%
  gl.filter.callrate(method = "ind", threshold = 0.754) %>%
  gl.filter.maf(threshold = 0.0074) %>%
  gl.filter.secondaries(method = "best") %>%
  gl.filter.reproducibility(threshold = 0.96) %>%
  gl.filter.monomorphs() %>%
  gl.recalc.metrics()
```

***Explore filtered data***
```{r}
#Explore the missingness
gl.report.callrate(filtered_gl)
```

***Export to vcf and import back to R***
```{r}
gl2vcf(filtered_gl, outfile = "gl.vcf",
       snp_pos = "ChromPosSnp_Sorghum_bicolor_v5.1",
       snp_chr = "Chrom_Sorghum_bicolor_v5.1")
```





***Parse data into seed companies***
```{r}
Rosenow <- gl.drop.pop(filtered_gl, pop.list = c("Karper", "Quinby"), recalc = TRUE, verbose = NULL) %>%
  gl.recalc.metrics()
Quinby <- gl.drop.pop(filtered_gl, pop.list = c("Rosenow", "Karper"), recalc = TRUE, verbose = NULL) %>%
  gl.recalc.metrics()
Karper <- gl.drop.pop(filtered_gl, pop.list = c("Rosenow", "Quinby"), recalc = TRUE, verbose = NULL) %>%
  gl.recalc.metrics()
```

***Neighbour joining analysis***
```{r}
#Export seed companies to fasta format
gl2fasta(Rosenow, method = 1, outfile = "Rosenow.fasta",
         outpath = getwd())
gl2fasta(Franklin, method = 1, outfile = "Franklin.fasta",
         outpath = getwd())
gl2fasta(Quinby, method = 1, outfile = "Quinby.fasta",
         outpath = getwd())
rm(list = ls())

```

*Import fasta to r for phylogentic tree and plot out*
```{r}
Rosenow <- fasta2DNAbin(file = "Rosenow.fasta")
D <- dist.dna(Rosenow,model = "K80")
tre <- nj(D)
#class(tre)
tre <- ladderize(tre)
#plot(tre, cex = .6)
C <- hclust(D, method = "average", members = NULL)
plot(C, cex = .6, main = "")
title("Rosenow (filtered)")

Franklin <- fasta2DNAbin(file = "Franklin.fasta")
D <- dist.dna(Franklin,model = "K80")
tre <- nj(D)
#class(tre)
tre <- ladderize(tre)
#plot(tre, cex = .6)
C <- hclust(D, method = "average", members = NULL)
plot(C, cex = .6, main = "")
title("Franklin (filtered)")

Quinby <- fasta2DNAbin(file = "Quinby.fasta")
D <- dist.dna(Quinby,model = "K80")
tre <- nj(D)
#class(tre)
tre <- ladderize(tre)
#plot(tre, cex = .6)
C <- hclust(D, method = "average", members = NULL)
plot(C, cex = .6, main = "")
title("Quinby (filtered)")

rm(list = ls()) #clear environment
```

***Skip filtration and conduct Neighbour Joining Analysis***
```{r}
#less stringent filtering
gl <- gl %>%
  gl.filter.monomorphs() %>%
  gl.recalc.metrics()

#Parse data by population
Rosenow2 <- gl.drop.pop(gl, pop.list = c("Karper", "Quinby"), recalc = TRUE) %>%
  gl.recalc.metrics()
Quinby2 <- gl.drop.pop(gl, pop.list = c("Rosenow", "Karper"), recalc = TRUE) %>%
  gl.recalc.metrics()
Karper2 <- gl.drop.pop(gl, pop.list = c("Rosenow", "Quinby"), recalc = TRUE)

#Rename sample ids in Quinby
#Quinby2 <- Quinby2 %>%
#  gl.recode.ind(ind.recode = "../Quinby_rename.csv", recalc = TRUE)

#Export genlight to a fasta format
gl2fasta(Rosenow2, method = 1, outfile = "Rosenow2.fasta",
         outpath = getwd())
gl2fasta(Karper2, method = 1, outfile = "Karper2.fasta",
         outpath = getwd())
gl2fasta(Quinby2, method = 1, outfile = "Quinby2.fasta",
         outpath = getwd())
rm(list = ls())

#jpeg(file = "my_plot.pdf", width = 18,height = 15)

Rosenow <- fasta2DNAbin(file = "Rosenow2.fasta")
D <- dist.dna(Rosenow,model = "K80")
tre <- nj(D)
#class(tre)
tre <- ladderize(tre)
#plot(tre, cex = .6)
C <- hclust(D, method = "average", members = NULL)
plot(C, cex = .6, main = "")
title("Rosenow (unfiltered)")

Karper <- fasta2DNAbin(file = "Karper2.fasta")
D <- dist.dna(Karper,model = "K80")
tre <- nj(D)
#class(tre)
tre <- ladderize(tre)
#plot(tre, cex = .6)
C <- hclust(D, method = "average", members = NULL)
plot(C, cex = .6, main = "")
title("Franklin (unfiltered)")

Quinby <- fasta2DNAbin(file = "Quinby2.fasta")
D <- dist.dna(Quinby,model = "K80")
tre <- nj(D)
#class(tre)
tre <- ladderize(tre)
#plot(tre, cex = .6)
C <- hclust(D, method = "average", members = NULL)
plot(C, cex = .6, main = "")
title("Quinby (unfiltered)")


```

***PCA***
```{r}
pca <- gl.pcoa(filtered_gl) #generate the pca
gl.pcoa.plot(pca, filtered_gl) #plot pca
```

```{r}
rose_pca <- gl.pcoa(Rosenow)
gl.pcoa.plot(rose_pca, Rosenow)
```
```{r}
quin_pca <- gl.pcoa(Quinby)
gl.pcoa.plot(quin_pca, Quinby)
```



```{r}
karp_pca <- gl.pcoa(Franklin)
gl.pcoa.plot(karp_pca, Franklin, ellipse = TRUE)
```

```{r}
pop(filtered_gl) <- filtered_gl@other$ind.metrics$Generation
gl_pca <- gl.pcoa(filtered_gl)
gl.pcoa.plot(gl_pca, filtered_gl, ellipse = TRUE)

```

