---
title: "Playing Around with Fitness Functions"
output: html_notebook
---

Imports
```{r}
library(AlphaSimR)
library(ggplot2)
library(plotly)

rm(list = ls())
```

Functions
```{r}
plot_fitness <- function(df) {
  ggplot(data=df, aes(x=gen, y=fitness)) +
    geom_line() +
    geom_point() +
    labs(x = "Generation", y = "Fitness")
}

plot_hist <- function(pop) {
  gv_a = gv(pop)[,1]
  gv_b = gv(pop)[,2]
  idx = c(1:length(gv_a))
  df <- data.frame(gv_a, gv_b)
  df <- melt(as.data.table(df))
  ggplot(df, aes(x=value, color=variable)) + geom_histogram(binwidth=1, position='identity')
}

fitnessFunc <- function(x) {
  return (-(x)^2) + rnorm(1,sd=2)
}
```

Variables
```{r}
nGens <- 100
nInd <- 1000
nSegSites <- 1000
nChr <- 10
```

Show that all individuals converge to a fitness maximum
```{r}
start_trait_val <- c(-50,-25,-10,10,25,50)
numSims <- length(start_trait_val)

df <- data.frame(gen=rep(1:nGens, times=numSims),
                 fitness=numeric(nGens*numSims),
                 initialVal = numeric(nGens*numSims),
                 traitVal=numeric(nGens*numSims))

for (s in 1:numSims) {
  initVal <- start_trait_val[s]
  founders = quickHaplo(
     nInd=nInd,
     nChr=nChr,
     segSites=nSegSites
  )
  SP <- SimParam$new(founders)
  SP$addTraitA(
    10,
    mean=initVal,
    var=abs(initVal/2)
  )
  SP$setVarE(H2=0.5)
  pop <- newPop(founders, simParam=SP)
  
  for(gen in 1:nGens) {
    idx = (s-1)*nGens + gen
    df$fitness[idx] <- fitnessFunc(meanG(pop))
    df$traitVal[idx] <- meanG(pop)
    df$initialVal[idx] <- initVal
    pop <- selectInd(pop=pop, trait=fitnessFunc, nInd=nInd/2)
    pop <- randCross(pop, nCrosses=nInd)
  }
}

df$initialVal = as.character(df$initialVal)
ggplot(data=df, aes(x=gen, y=traitVal)) +
  geom_line(aes(color = initialVal)) +
  labs(x = "Generation", y = "Trait Value")

```

# Plot Trait Value against Fitness
```{r}
founders = quickHaplo(
   nInd=nInd,
   nChr=nChr,
   segSites=nSegSites
)
SP <- SimParam$new(founders)
SP$addTraitA(
  10,
  mean=50,
  var=abs(initVal/2)
)
SP$setVarE(H2=0.5)
pop <- newPop(founders, simParam=SP)

fit_df <- data.frame(gen=1:nGens,
                 fitness=numeric(nGens),
                 traitVal=numeric(nGens))
for(gen in 1:nGens) {
  fit_df$fitness[gen] <- fitnessFunc(meanG(pop))
  fit_df$traitVal[gen] <- meanG(pop)
  pop <- selectInd(pop=pop, trait=optFit, nInd=nInd/2)
  pop <- randCross(pop, nCrosses=nInd)
}

par(mar = c(5, 5, 3, 5) + 0.3)
plot(fit_df$gen, fit_df$traitVal, type="l", lwd = "3", col=2, xlab = "Generation", ylab = "Trait Value")   
par(new = TRUE) 
plot(fit_df$gen, fit_df$fitness, type="l", lwd = "3", col = 3, axes = FALSE, xlab = "", ylab = "") 
axis(side = 4, at = pretty(range(fit_df$fitness)))
mtext("Fitness", side = 4, line = 3)
par(xpd=TRUE)
legend("right",
  c("Trait Value", "Fitness"),
       lty = 1,
       lwd = 3,
       col = 2:3)
```


Play around with different fitness functions
```{r}
quad <- function(x, optima)
  return (-(x-optima)^2)

x = seq(-100,100, by=1)
y = quad(x, 0)

df <- data.frame(x=x, y=y)
ggplot(data=df, aes(x=x, y=y)) +
    geom_line() +
    geom_point() +
    labs(x = "Trait Value", y = "Fitness")
```


Plot a 3d surface of GEBV against 2 different input parameters

Variables to play around with:
- Selection intensity
- Population size
- Heritability
- Starting genetic variance
```{r}
pop_vec <- seq(100,2100, by=400)
sel_int_vec <- seq(0.1,0.9, by=0.2)
H2 <- 0.5
nSims <- 3
nGens <- 100
nSegSites <- 1000
nChromosomes <- 10
nParents <- 1000
nQtl <- nSegSites/nChromosomes
gvar <- 10

genMat <- matrix(NA, nrow=length(pop_vec),ncol=length(sel_int_vec))
rownames(genMat) <- pop_vec
colnames(genMat) <- sel_int_vec

for (p in 1:length(pop_vec)) {
  popSize = pop_vec[p]
  print(paste0("Pop Size: ", popSize))
  for (s in 1:length(sel_int_vec)) {
    selInt <- sel_int_vec[s]
    gain <- vector(length=nSims)
    for (sim in c(1:nSims)) {
      founders = quickHaplo(
        nInd=popSize,
        nChr=nChromosomes,
        segSites=nSegSites
      )
      SP <- SimParam$new(founders)
    
      SP$addTraitA(
        mean = 0,
        var = gvar,
        nQtlPerChr = nQtl
      )
      SP$setVarE(H2=H2)
      pop <- newPop(founders, simParam=SP)
      
      startGV <- meanG(pop)
      for(gen in 1:nGens) {
        df$traitVal[gen] <- meanG(pop)
        # Adjust this to use the fitness function instead of phenotype
        pop <- selectCross(pop=pop, use='pheno', nInd=popSize*selInt, nCrosses = popSize)
      }
      gain[sim] <- meanG(pop) - startGV
    }
    genMat[p, s] <- median(gain)
  }
}

# Use this to plot a 3d surface
# The input matrix z should have y rows and x columns
fig <- plot_ly(x=sel_int_vec, y=pop_vec, z = genMat, type='surface') %>%
  layout(scene = list(xaxis = list(title = "Selection Intensity"),
                      yaxis = list(title = "Population Size"),
                      zaxis = list(title = "Genetic Gain"),
                      aspectmode='cube'))
fig
```

Plot Genetic Gain as a Function of Population Size
```{r}
parent_vec <- c(100,8000)
nGens <- 200
nSegSites <- 200
nChromosomes <- 10
nQtl <- nSegSites/nChromosomes
selectionIntensity <- 0.8
gvar <- 5
her = 0.5

nCombs = length(parent_vec)
df <- data.frame(gen=rep(1:nGens, times=nCombs),
                 traitVal=numeric(nGens*nCombs),
                 popSize=numeric(nGens*nCombs))

for (p in 1:length(parent_vec)) {
  nParents = parent_vec[p]
  print(paste0("Parents: ", nParents))
  
  founders = quickHaplo(
    nInd=nParents,
    nChr=nChromosomes,
    segSites=nSegSites
  )
  SP <- SimParam$new(founders)

  SP$addTraitA(
    mean = 0,
    var = gvar,
    nQtlPerChr = nQtl
  )
  SP$setVarE(H2=her)
  pop <- newPop(founders, simParam=SP)
  
  startGV <- meanG(pop)
  for(gen in 1:nGens) {
    idx = (p-1)*nGens + gen
    df$traitVal[idx] <- meanG(pop)
    df$popSize[idx] <- nParents
    pop <- selectCross(pop=pop, use='pheno', nInd=nParents*selectionIntensity, nCrosses = nParents)
  }
}

df$popSize = as.character(df$popSize)
ggplot(data=df, aes(x=gen, y=traitVal)) +
  geom_line(aes(color = popSize)) +
  labs(x = "Generation", y = "GEBV")
```


[Work in progress]
Out of alleles left in the population, how many of them would increase fitness?
Also: how does this change with size of allele? Look at editGenomeTopQtl()
Also: how does size of 'fitness delta' change over generations?

# run for multiple individuals, and pull out individuals and cross them
# inspect genetic architecture
# show that P(allelic substitution is favorable)
```{r}


hetLocus <- function(locus) {
  return (length(unique(locus)) > 1)
}

initVal = 10
popSize = 1000
nChr = 10
nSegSites = 100
nLoci = nSegSites * nChr
nGens = 20

founders = runMacs(
   nInd=popSize,
   nChr=nChr,
   segSites=nSegSites
)
SP <- SimParam$new(founders)
SP$addTraitA(
  10,
  mean=initVal,
  var=abs(initVal/2)
)

SP$setVarE(H2=0.5)
pop <- newPop(founders, simParam = SP)

cols = colnames(pullSegSiteGeno(pop))
df <- data.frame(gen=rep(1:nGens),
                 fitness=numeric(nGens),
                 traitVal=numeric(nGens),
                 percFitInc=numeric(nGens))

for (gen in 1:nGens) {
  print(gen)
  df$fitness[gen] <- fitnessFunc(meanG(pop))
  df$traitVal[gen] <- meanG(pop)
  num_het_loci = 0
  num_inc = 0
  geno <- pullSegSiteGeno(pop)
  for (l in 1:nLoci) {
    strs = unlist(strsplit(cols[l], "_"))
    chr = strtoi(strs[1])
    site = strtoi(strs[2])
    locus = geno[,l]
    if (hetLocus(locus)) {
      num_het_loci <- num_het_loci + 1
      inc <- FALSE
      if (FALSE) {
        for (i in 1:nInd(pop)) {
          ind = geno[i,]
          fit_pre <- fitnessFunc(pop@gv[i])
          pop0 <- editGenome(pop, ind=i, chr=chr, segSites=site, allele=0, simParam=SP)
          pop1 <- editGenome(pop, ind=i, chr=chr, segSites=site, allele=1, simParam=SP)
          
          fit_post0 <- fitnessFunc(pop0@gv[i])
          fit_post1 <- fitnessFunc(pop1@gv[i])
          if ((fit_post0 > fit_pre) || (fit_post1 > fit_pre) ) {
            inc <- TRUE
            break
          }
        }
        if (inc) {
          num_inc <- num_inc + 1
        }
      }
      
    }
  }
  
  if (num_het_loci == 0) {
    perc <- 0
  } else {
    perc <- (num_inc / num_het_loci)
  }
  df$percFitInc[gen] <- num_het_loci
  pop <- selectCross(pop=pop, trait=fitnessFunc, nInd=popSize/2, nCrosses=popSize)
}

par(mar = c(5, 5, 3, 5) + 0.3)
plot(df$gen, df$traitVal, type="l", lwd = "3", col=2, xlab = "Generation", ylab = "Trait Value")
par(new = TRUE) 
plot(df$gen, df$percFitInc, type="l", lwd = "3", col = 3, axes = FALSE, xlab = "", ylab = "") 
axis(side = 4, at = pretty(range(df$fitness)))
mtext("% of Segregating alleles that are Beneficial", side = 4, line = 3)
par(xpd=TRUE)
legend("right",
  c("Trait Value", "% Alleles"),
       lty = 1,
       lwd = 3,
       col = 2:3)

ggplot(data=df, aes(x=gen, y=percFitInc)) +
  geom_line() +
  labs(x = "Generation", y = "% Segregating Alleles that are Beneficial")

```
